import-scripts/add-age-at-seq-report.py

```diff
@@ -1,55 +1,49 @@
+#!/usr/bin/env python
+
 import sys
 import optparse
 import csv
 import os
 import math
-from datetime import datetime
 
 ERROR_FILE = sys.stderr
 OUTPUT_FILE = sys.stdout
 
 # globals
 AGE_AT_SEQ_REPORT_FIELD = 'AGE_AT_SEQ_REPORT'
-AGE_FIELD = 'AGE'
-SEQ_DATE_FIELD = 'SEQ_DATE'
+AGE_AT_SEQ_REPORTED_YEARS_FIELD = 'AGE_AT_SEQ_REPORTED_YEARS'
 AVERAGE_DAYS_PER_YEAR = 365.2422
 
 PATIENT_SAMPLE_MAP = {}
-SAMPLE_SEQ_DATE_MAP = {}
-PATIENT_AGE = {}
+SAMPLE_AGE_AT_SEQ_REPORT_MAP = {}
 
 
-def load_sample_seq_date(seq_date_file, convert_to_days):
-	""" Loads SEQ_DATE from seq date file provided. """
-	data_file = open(seq_date_file, 'rU')
-	data_reader = csv.DictReader(data_file, dialect = 'excel-tab')
+def read_sample_file(sample_file):
+    """ Generator function that yields all non-commented lines from the clinical sample file. """
+    with open(sample_file, 'rU') as f:
+        for line in f:
+            if line.startswith('#'):
+                continue
+            yield line
+
+
+def load_age_at_seq_reported_years(sample_file, convert_to_days):
+	""" Loads AGE_AT_SEQ_REPORTED_YEARS from clinical sample file and converts it to days. """
 
+	data_reader = csv.DictReader(read_sample_file(sample_file), dialect = 'excel-tab')
 	for line in data_reader:
+		sample_id = line['SAMPLE_ID']
 		# skip samples not found in the clinical file
-		if not line['SAMPLE_ID'] in PATIENT_SAMPLE_MAP.keys():
+		if not sample_id in PATIENT_SAMPLE_MAP.keys():
 			continue
-		seq_date = datetime.strptime(line['SEQ_DATE'], '%a, %d %b %Y %H:%M:%S %Z')
-		difference_in_years = int(math.ceil(((datetime.now() - seq_date).days)/AVERAGE_DAYS_PER_YEAR))
+		age_at_seq_reported_years = line[AGE_AT_SEQ_REPORTED_YEARS_FIELD]
 		try:
-			age_at_seq_report_value = (PATIENT_AGE[line['PATIENT_ID']] - difference_in_years)
 			if convert_to_days:
-				age_at_seq_report_value = age_at_seq_report_value*AVERAGE_DAYS_PER_YEAR
-			SAMPLE_SEQ_DATE_MAP[line['SAMPLE_ID']] = str(int(math.ceil(age_at_seq_report_value)))
-		except KeyError:
-			print "Patient age not found for '" + line['PATIENT_ID'] + "'"
-			SAMPLE_SEQ_DATE_MAP[line['SAMPLE_ID']] = 'NA'
-
-
-def load_patient_age(age_file):
-	""" Loads AGE from age file. """
-	data_file = open(age_file, 'rU')
-	data_reader = csv.DictReader(data_file, dialect = 'excel-tab')
-
-	for line in data_reader:
-		# skip records not found in clinical file 
-		if not line['PATIENT_ID'] in PATIENT_SAMPLE_MAP.values():
-			continue
-		PATIENT_AGE[line['PATIENT_ID']] = int(line['AGE'])
+				age_at_seq_report_value = int(age_at_seq_reported_years) * AVERAGE_DAYS_PER_YEAR
+			SAMPLE_AGE_AT_SEQ_REPORT_MAP[sample_id] = str(int(math.floor(age_at_seq_report_value)))
```

Comment:
I notice that ceil() was changed to floor() here. That may be fine ... so long as the same calculation is being performed consistently (for age at sequencing and age at death)

Reply:
Thanks for noticing this @sheridancbio! I changed it from `ceil()` to `floor()` because it seemed to me that rounding up this value would introduce the same kind of errors that we were seeing before. When I run this same code but with the `math.ceil()` function, I see 444 records where the patient's `age_at_sequencing_reported` is greater than `int_dod`, and in a lot of these cases `AGE_AT_SEQUENCING_REPORTED` is only greater by one day. When I run it with `math.floor()`, I only see 25 records where `AGE_AT_SEQUENCING_REPORTED` is greater than `INT_DOD`. I guess it depends what we want. I'm not sure how this value is being converted to years in the genie portal - ie if they're dividing by 365 or 365.2422 (the multiplier we're using). It could yield different results based on whether we're using `floor()` or `ceil()`.

To your second point - that we need the calculations to be performed consistently - I don't think this would be possible without a larger code change. Age at death (INT_DOD) is calculated [here](https://github.com/knowledgesystems/cmo-pipelines/blob/72c387529b737952fba0d551a5ec0bb9f7d58149/ddp/ddp_pipeline/src/main/java/org/mskcc/cmo/ks/ddp/pipeline/model/SuppVitalStatusRecord.java#L63), where it calls [this function](https://github.com/knowledgesystems/cmo-pipelines/blob/4ea50c11a13b18e5601ec53877a9c634715fe31c/ddp/ddp_pipeline/src/main/java/org/mskcc/cmo/ks/ddp/pipeline/util/DDPUtils.java#L208-L216). From reading the code it seems that `INT_DOD` is precise to the day (and might be revealing the patient's birthdate?) Maybe we could discuss this a bit?

Reply:
I guess it is a little confusing and philosophical to decide exactly when a person changes from being X years old to being X + 1 years old. Do you become a year older at midnight of the day of your birth, at midnight of the day after your birth, or at your time of birth on the day of your birth? Even the last option is debatable because a year of time (astronomically) includes a fraction of a day, and that fraction is what we are either rounding up or down here in an aggregate over a person's lifetime. I think your comment about knowing the way that genie reverses this conversion is on point - it does seem to me that it would be improper for us to send to genie a number of days for a patient whom we consider to be 30 ... and have genie integrate that patient into their dataset with an age of 31. It seems to me that whatever (year) age we consider the patient to be we should convert that to days in a way where the reverse conversion by genie will always result in the original age we had.

If our age at death calculation, I see the same conversion factor here also present in resolvePatientCurrentAge(), but the reference you give points to resolveIntervalInDays() and that only does anonymization of age based on upper/lower limit rather than on reducing the granularity to years. So I think something is missing here ... the granularity of the code you pointed me to is days (I think) .. but isn't age at death recorded in years in our clinical file? There must be another step where we convert the days into years, and that would be the logic which should ideally be consistent I guess.

Reply:
@sheridancbio Here's the code where genie converts the age at seq in days to years: https://github.com/Sage-Bionetworks/Genie/blob/53db39ccb6e423743df49c1f0e115c9a7576af69/genie/database_to_staging.py#L1132-L1137. I've modified our code to do the reverse calculation. I'll note that this increases the number of cases where age of sequencing is greater than INT_DOD to 636 records but again it seems that the majority of these cases are off by 1 day, and now that we're doing the exact reverse calculation they should end up with the correct year value.